crates/sui-core/src/transaction_manager.rs

```diff
@@ -28,7 +28,10 @@ pub(crate) struct TransactionManager {
 #[derive(Default)]
 struct Inner {
     // Maps missing input objects to transactions in pending_certificates.
-    missing_inputs: BTreeMap<ObjectKey, TransactionDigest>,
+    // Note that except for immutable objects, a given key may only have one TransactionDigest in
```

Comment:
I like the current approach since it is the simplest. But for immutable vs mutable objects, can we in theory propagate the property out of `get_missing_input_objects()`?https://github.com/mwtian/sui/blob/6788f83b08be0e9c73d986b83d2088752c7378b0/crates/sui-core/src/authority/authority_store.rs#L331

Reply:
I tried to do that, but you can't. The case you linked to is where the object exists, but we have to further check if the lock exists. The immutability is only stored in the object metadata, so we can't do anything useful in the case where the object doesn't exist at all. I may investigate splitting InputObjectKind::ImmOrOwnedMoveObject into two enums to allow this.